docs/src/main/asciidoc/panache-rest.adoc

```diff
@@ -0,0 +1,242 @@
+////
+This guide is maintained in the main Quarkus repository
+and pull requests should be submitted there:
+https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
+////
+
+= Quarkus - Generating CRUD resources with Panache REST
+
+include::./attributes.adoc[]
+:extension-status: experimental
+
+A lot of web applications are monotonous CRUD applications with REST APIs that are tedious to write.
+To streamline this task, Panache REST can generate the basic CRUD endpoints for your entities and repositories.
+
+While this extension is still experimental and provides a limited feature set, we hope to get an early feedback for it.
+Currently this extension supports Hibernate ORM with Panache and can generate CRUD resources that work with `application/json` and `application/hal+json` content.
+
+include::./status-include.adoc[]
+
+== Setting up Panache REST
+
+To begin with:
+
+* Add the required dependencies to your `pom.xml`
+** Panache REST with Hiberante ORM extension (`quarkus-panache-rest-hibernate-orm`)
+** A JDBC driver extension (`quarkus-jdbc-postgresql`, `quarkus-jdbc-h2`, `quarkus-jdbc-mariadb`, ...)
+** One of the RestEasy JSON serialization extensions (`quarkus-resteasy-jackson` or `quarkus-resteasy-jsonb`)
+
+[source,xml]
+----
+<dependencies>
+    <!-- Panache REST with Hibernate ORM -->
+    <dependency>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-panache-rest-hibernate-orm</artifactId>
+    </dependency>
+
+    <!-- JDBC driver dependencies -->
+    <dependency>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-jdbc-postgresql</artifactId>
+    </dependency>
+
+    <!-- JSON serializer dependency -->
+    <dependency>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-resteasy-jackson</artifactId>
+    </dependency>
+</dependencies>
+----
+
+* Implement the Panache entities and/or repositories as explained in the link:hibernate-orm-panache[Panache with Hibernate ORM guide].
+* Define the interfaces for generation as explained in the sections below.
+
+== Generating CRUD resources
+
+Panache REST generates CRUD resources based on the interfaces available in your application.
+For each entity and repository that you want to generate, provide an interface that extends either `PanacheEntityCrudResource` or `PanacheRepositoryCrudResource` interface.
+_Do not implement these interfaces and don't provide custom methods because they will be ignored._ You can, however, override the methods from the extended interface in order to customize them (see the section at the end).
+
+The generated resources will be functionally equivalent for both entities and repositories.
+The only difference being the particular data access pattern in use.
+
+Panache REST will generate a JAX-RS resource class similar to this:
+
+[source,java]
+----
+@Path("/entities") // Base bath is a pluralised name of an entity
```

Comment:
I think this is a mistake. We shouldn't "pluralize" things as there's a good chance we will get some of them wrong and it really doesn't bring anything.

Keep in mind that a lot of people code in their own languages where the plural rules will be completely different.

Also, frankly, I don't think it's better as a plural so why take the risk?

Reply:
/cc @FroMage

Reply:
If we want plural, perhaps let the user specify it?

Reply:
Just to give an example, in French, we have a lot of specific plural forms (`cheval` -> `chevaux` for instance).

Reply:
Yeah, taking the entity name as is would be the default and you could define a specific path in the annotation?

Reply:
Would you mind to create an issue for this and assign to me? I'll need to change the extension because it now pluralises everything.
I'll also add a way to override it separately.

Just an FIY Spring Data REST simply adds `s` to every entity name.

Reply:
> Just an FIY Spring Data REST simply adds s to every entity name.

Oh, my. OK I will create an issue.

Reply:
Makes it even more fun with:

```java
@Entity
public class 馬 extends PanacheEntity {
}
```

And the URI ends up being `/馬s`, right? :)

Myself, I've always used `Cheval` for the entity name (singular) and `Chevaux` (plural) for the controller name. And RoR derives the URIs from the controller names, so it would be `/chevaux` correctly.

I just suggest we advertise the same convention, it solves many issues. And we default to the controller name for the URI, instead of the entity name.

Naturally we should also allow users to override it.

Reply:
> And the URI ends up being `/馬s`, right? :)

I would think so. At least for a simple `people` I got a path `/peoples` when i tried it last time.

> I just suggest we advertise the same convention, it solves many issues. And we default to the controller name for the URI, instead of the entity name.

Extracting a name from a controller will be trickier because a user can name his classes differently. We could recommend a convention but who knows if they'll follow it.
With an entity we could simply take an its name, make it all lowercase separated with `-` instead of camelcase and get on with it. And then I'll add an option to override the path.

Reply:
> Extracting a name from a controller will be trickier because a user can name his classes differently. We could recommend a convention but who knows if they'll follow it.

The point of conventions is that people follow them. If they don't they can't customise the output.

Entities are already singular by convention. Controller names don't have a convention other than the brain-dead "resource" that jax-rs got confused about.

It's pretty easy to tell people to name their controller `People` and default to using that name.

Reply:
Let's discuss it on https://github.com/quarkusio/quarkus/issues/9261 since the guide change won't be enough for this anyway.